x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java

```diff
@@ -7,55 +7,212 @@
+            fields.forEach(alias -> {
+                // TODO: can we support CARTESIAN also?
+                if (alias.child() instanceof StDistance distance && distance.crsType() == BinarySpatialFunction.SpatialCrsType.GEO) {
+                    distances.put(alias.id(), distance);
+                } else if (alias.child() instanceof Attribute attr) {
+                    aliasReplacedByBuilder.put(alias.toAttribute(), attr.toAttribute());
+                }
+            });
+            AttributeMap<Attribute> aliasReplacedBy = aliasReplacedByBuilder.build();
+
+            List<EsQueryExec.Sort> pushableSorts = new ArrayList<>();
+            for (Order order : orders) {
+                if (LucenePushDownUtils.isPushableFieldAttribute(order.child(), hasIdenticalDelegate)) {
+                    pushableSorts.add(
+                        new EsQueryExec.FieldSort(
+                            ((FieldAttribute) order.child()).exactAttribute(),
+                            order.direction(),
+                            order.nullsPosition()
+                        )
+                    );
+                } else if (order.child() instanceof ReferenceAttribute referenceAttribute) {
+                    if (distances.containsKey(referenceAttribute.id())) {
+                        StDistance distance = distances.get(referenceAttribute.id());
+                        StDistance d = (StDistance) distance.transformDown(ReferenceAttribute.class, r -> aliasReplacedBy.resolve(r, r));
+                        PushableGeoDistance pushableGeoDistance = PushableGeoDistance.from(d, order);
+                        if (pushableGeoDistance != null) {
+                            pushableSorts.add(pushableGeoDistance.sort());
+                        } else {
+                            // As soon as we see a non-pushable sort, we know we need a final SORT command
+                            break;
+                        }
+                    } else if (aliasReplacedBy.resolve(referenceAttribute, referenceAttribute) instanceof FieldAttribute fieldAttribute
+                        && LucenePushDownUtils.isPushableFieldAttribute(fieldAttribute, hasIdenticalDelegate)) {
+                            // If the SORT refers to a reference to a pushable field, we can push it down
+                            pushableSorts.add(
+                                new EsQueryExec.FieldSort(fieldAttribute.exactAttribute(), order.direction(), order.nullsPosition())
+                            );
+                        } else {
+                            // If the SORT refers to a non-pushable reference function, the EVAL must remain before the SORT,
+                            // and we can no longer push down anything
+                            break;
+                        }
+                } else {
+                    // As soon as we see a non-pushable sort, we know we need a final SORT command
+                    break;
```

Comment:
Bug! I think in these two cases, we should `return NO_OP` instead of `break`ing - in case we find one pushable sort but then the next sort is not pushable.

Reproducer:
```
FROM test | sort st_distance(x, \"POINT(0 0)\"::geo_point) asc, x::keyword desc
```
This pushes down the distance sort but ignores the sort on `x::keyword`:
```
[2024-10-09T11:36:07,239][INFO ][o.e.x.e.o.LocalPhysicalPlanOptimizer] [runTask-0] Rule local.PushTopNToSource applied
ProjectExec[[x{f}#20, $$order_by$0$1{r}#21, $$order_by$1$2{r}#22]]                                                          = ProjectExec[[x{f}#20, $$order_by$0$1{r}#21, $$order_by$1$2{r}#22]]
\_TopNExec[[Order[$$order_by$0$1{r}#21,ASC,LAST], Order[$$order_by$1$2{r}#22,DESC,FIRST]],1000[INTEGER],null]               ! \_EvalExec[[STDISTANCE(x{f}#20,[1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0][GEO_POINT]) AS $$order_by$0$1, TOSTRING(x{f}#2
  \_EvalExec[[STDISTANCE(x{f}#20,[1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0][GEO_POINT]) AS $$order_by$0$1, TOSTRING(x{f}#2 ! 0) AS $$order_by$1$2]]
0) AS $$order_by$1$2]]                                                                                                      !   \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#23], limit[1000], sort[[GeoDistanceSort[field=x{f}#20, direction=ASC, lat=0.0, lon=0.0]]] estimatedRowSize[null]
    \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#23], limit[], sort[] estimatedRowSize[null]                   ! 

```